src/views/product-root-docs-path-landing/server.ts

```diff
@@ -0,0 +1,135 @@
+import { RootDocsPath } from 'types/products'
+import { getStaticGenerationFunctions as _getStaticGenerationFunctions } from 'layouts/sidebar-sidecar/server'
+import { GenerateGetStaticPropsArguments } from './types'
+
+/**
+ * @TODO add TS to function signature & document function purpose
+ */
+const generateHeadingLevelsAndSidecarHeadings = ({
+  layoutHeadings,
+  marketingContentBlocks,
+  pageTitle,
+}) => {
+  const additionalHeadings = []
+
+  let currentSectionHeading
+  const marketingContentBlocksWithHeadingLevels = marketingContentBlocks.map(
+    (block) => {
+      const blockCopy = { ...block }
+
+      let thisHeadingObject
+
+      if (block.type === 'section-heading') {
+        // all section-heading block types are supposed to be h2's
+        const headingSlug = slugify(block.title, { lower: true })
+        thisHeadingObject = {
+          level: 2,
+          title: block.title,
+          id: headingSlug,
+          slug: headingSlug,
+        }
+        currentSectionHeading = thisHeadingObject
+      } else if (block.type === 'card-grid') {
+        // all card-grid headings will be h3's unless a section-heading is before it
+        let cardGridHeadingLevel
+        if (currentSectionHeading) {
+          cardGridHeadingLevel = currentSectionHeading.level + 1
+        } else {
+          cardGridHeadingLevel = 2
+        }
+
+        const headingSlug = slugify(block.title, { lower: true })
+        thisHeadingObject = {
+          level: cardGridHeadingLevel,
+          title: block.title,
+          id: headingSlug,
+          slug: headingSlug,
+        }
+      }
+
+      if (thisHeadingObject) {
+        additionalHeadings.push(thisHeadingObject)
+        blockCopy.headingLevel = thisHeadingObject.level
+        blockCopy.headingId = thisHeadingObject.id
+      }
+
+      return blockCopy
+    }
+  )
+
+  const sidecarHeadings = [
+    ...layoutHeadings.map((heading) => {
+      if (heading.level === 1) {
+        const slug = slugify(pageTitle, { lower: true })
+        return { level: 1, id: slug, slug, title: pageTitle }
+      } else {
+        return heading
+      }
+    }),
+    ...additionalHeadings,
+  ]
+
+  return { sidecarHeadings, marketingContentBlocksWithHeadingLevels }
+}
+
+const generateGetStaticProps = ({
+  includeMDXSource = false,
+  pageContent,
+  product,
+}: GenerateGetStaticPropsArguments) => {
+  const basePath = 'docs'
+  const currentRootDocsPath = product.rootDocsPaths.find(
+    (rootDocsPath: RootDocsPath) => rootDocsPath.path === basePath
+  )
+  const baseName = currentRootDocsPath.shortName
+
+  return async (context: GetStaticPropsContext) => {
+    const { getStaticProps: generatedGetStaticProps } =
+      _getStaticGenerationFunctions({
+        product,
+        basePath,
+        baseName,
+      })
+
+    // TODO: replace any with accurate type
+    const generatedProps = (await generatedGetStaticProps({
+      ...context,
+      params: { page: [] },
+    })) as any
```

Comment:
⛏️ Nitpick: Not even sure if this is preferable, but we could do `$TSFixMe` here for track-ability:

```suggestion
    })) as $TSFixMe
```

Reply:
Oh, I forgot about this! 🙌🏻 I'll make this change in all the places applicable here!

Reply:
🗒️  Resolved in https://github.com/hashicorp/dev-portal/pull/452/commits/4b606fae06137ae51488b0e365f372932a4e2b18